hnn_qt5.py

```diff
@@ -218,71 +231,174 @@ def runsim (self):
     print(''); self.updatewaitsimwin('')
   
   def optmodel (self):
+    import simdat
+    simdat.initial_ddat = simdat.ddat.copy()
+
     self.updatewaitsimwin('Optimizing model. . .')
-    from neuron import h # for praxis
+
+    self.last_step = False
+    self.first_step = True
+    for step in dconf['opt_info'].keys():
+      self.opt_params = dconf['opt_info'][step]
+
+      self.cur_step = step
+      if self.cur_step == len(dconf['opt_info']) - 1:
+        self.last_step = True
+        # update ranges and recalculate weights
+        self.updateoptparams()
+        sleep(1)
+
+        # reload opt_params for the last step in case the number of
+        # steps was changed by updateoptparams()
+        self.opt_params = dconf['opt_info'][len(dconf['opt_info']) - 1]
+
+      print("Starting optimization step %d"%(step+1))
+      self.runOptStep()
+
+      if not self.best_ddat is None:
+        simdat.ddat = self.best_ddat.copy()
+
+      simdat.updateoptdat(paramf,simdat.ddat['dpl']) # update optdat with best from this step
+
+      # put best opt results into GUI
+      push_values = OrderedDict()
+      for param_name in self.opt_params['ranges'].keys():
+        push_values[param_name] = self.opt_params['ranges'][param_name]['initial']
+
+      self.updatebaseparamwin(push_values)
+      sleep(1)
+
+      if self.first_step:
+        # update the plots now because we didn't do it on the first iteration
+        # for the first step
+        self.updatedrawerr() # send event to draw updated error (asynchronously)
+        self.first_step = False
+
+    # one final sim with the best parameters to update display
+    self.runsim(is_opt=True)
+
+  def runOptStep (self):
+    import simdat
+
     self.optiter = 0 # optimization iteration
-    fpopt = open(os.path.join(dconf['datdir'],paramf.split(os.path.sep)[-1].split('.param')[0],'optinf.txt'),'w')
+    fnoptinf = os.path.join(dconf['datdir'],paramf.split(os.path.sep)[-1].split('.param')[0],'optinf.txt')
+    optinf_dir = os.path.dirname(fnoptinf)
+    os.makedirs(optinf_dir, exist_ok=True)
+    fpopt = open(fnoptinf,'w')
     fpopt.close()
     self.minopterr = 1e9
+    self.stepminopterr = self.minopterr
+    self.best_ddat = None
 
-    def optrun (vparam):
-      # create parameter dictionary of current values to test
-      lparam = list(dconf['params'].values())
+    def optrun (new_params, grad=0):
       dtest = OrderedDict() # parameter values to test      
-      for prm,val in zip(lparam,expvals(vparam,lparam)): # set parameters
-        if val >= prm.minval and val <= prm.maxval:
-          if debug: print('optrun prm:',prm.var,prm.minval,prm.maxval,val)
-          dtest[prm.var] = val
+      for param_name, test_value in zip(self.opt_params['ranges'].keys(), new_params): # set parameters
+        if test_value >= self.opt_params['ranges'][param_name]['minval'] and test_value <= self.opt_params['ranges'][param_name]['maxval']:
```

Comment:
This test is not strictly necessary with COBYLA, but in case the algorithm is changed at some point...